Runner.php

```diff
@@ -15,32 +15,65 @@ public function __construct($config, $server)
 
     public function queueDirectory($dir, $prefix = '')
     {
+        if(isset($this->config['include_paths'])) {
+            $this->queueWithIncludePaths();
+        } else {
+            $this->queuePaths($dir, $prefix);
+        }
+
+        $this->server->process_work(false);
+    }
+
+    public function queueWithIncludePaths() {
+        foreach ($this->config['include_paths'] as $f) {
+            if ($f !== '.' and $f !== '..') {
+
+                if (is_dir("/code$f")) {
+                    $this->queuePaths("/code$f", "$f/");
+                    continue;
+                }
+
+                if (isset($this->config['config']['file_extensions'])) {
+                    $this->filterByExtension($f);
+                } else {
+                    $this->server->addwork(array("/code/$f"));
+                }
+            }
+
+        }
+    }
+
+    public function queuePaths($dir, $prefix = '') {
         $dir = rtrim($dir, '\\/');
 
         foreach (scandir($dir) as $f) {
-            if (in_array("$prefix$f", $this->config["exclude_paths"])) {
+            if (!isset($this->config['include_paths']) && in_array("$prefix$f", $this->config["exclude_paths"])) {
```

Comment:
Really my design criticism is that it seems that `queuePaths` has too much responsibility. By this time we enter `queuePaths`, we've already decided whether we're using includes or excludes, so checking for them here again just makes the code harder to read.

So, my suggestion was, if you want to make `queuePaths` reusable, you could take exclusions in the argument list:

    public function queuePaths($dir, $prefix = '', $exclusions = []) {

And remove the explicit references to `$this->config` in line 50:

    if (in_array("$prefix$f", $exclusions)) {

Which means that you could use this for exclude-based file building this way:

    $this->queuePaths($dir, $prefix, $this->config['exclude_paths'])

But for include-based file building you could leave line 32 as:

    $this->queuePaths("/code$f", "$f/");

In that case, `$exclusions` will simply be an empty array, and will not skip any files in the directory. But you get the goal of reusing `queuePaths` while focusing its responsibilities.